### EDA

In [94]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from IPython.display import display #easier display in Editor

In [95]:
# Load the dataset
file1 = r'project_dataset/extract - fees - data analyst - .csv'
file2 = r'project_dataset/extract - cash request - data analyst.csv'
data1 = pd.read_csv(file1)
fees = data1.copy()
data2 = pd.read_csv(file2)
cash = data2.copy()

In [96]:
#exploring the dataset
display(fees.head(), cash.head())

id  cash_request_id             type    status               category  \
0   6537          14941.0  instant_payment  rejected                    NaN   
1   6961          11714.0         incident  accepted  rejected_direct_debit   
2  16296          23371.0  instant_payment  accepted                    NaN   
3  20775          26772.0  instant_payment  accepted                    NaN   
4  11242          19350.0  instant_payment  accepted                    NaN   

   total_amount                              reason  \
0           5.0  Instant Payment Cash Request 14941   
1           5.0               rejected direct debit   
2           5.0  Instant Payment Cash Request 23371   
3           5.0  Instant Payment Cash Request 26772   
4           5.0  Instant Payment Cash Request 19350   

                      created_at                     updated_at  \
0   2020-09-07 10:47:27.42315+00  2020-10-13 14:25:09.396112+00   
1  2020-09-09 20:51:17.998653+00  2020-10-13 14:25:15.537063+00   
2  2020-10-23 10:10:58.352972+00  2020-10-23 10:10:58.352994+00   
3  2020-10-31 15:46:53.643958+00  2020-10-31 15:46:53.643982+00   
4  2020-10-06 08:20:17.170432+00  2020-10-13 14:25:03.267983+00   

                         paid_at from_date to_date charge_moment  
0   2020-12-17 14:50:07.47011+00       NaN     NaN         after  
1   2020-12-08 17:13:10.45908+00       NaN     NaN         after  
2   2020-11-04 19:34:37.43291+00       NaN     NaN         after  
3  2020-11-19 05:09:22.500223+00       NaN     NaN         after  
4  2020-11-02 14:45:20.355598+00       NaN     NaN         after

id  amount    status                     created_at  \
0     5   100.0  rejected  2019-12-10 19:05:21.596873+00   
1    70   100.0  rejected   2019-12-10 19:50:12.34778+00   
2     7   100.0  rejected   2019-12-10 19:13:35.82546+00   
3    10    99.0  rejected  2019-12-10 19:16:10.880172+00   
4  1594   100.0  rejected  2020-05-06 09:59:38.877376+00   

                      updated_at  user_id                   moderated_at  \
0   2019-12-11 16:47:42.40783+00    804.0  2019-12-11 16:47:42.405646+00   
1  2019-12-11 14:24:22.900054+00    231.0  2019-12-11 14:24:22.897988+00   
2  2019-12-11 09:46:59.779773+00    191.0  2019-12-11 09:46:59.777728+00   
3  2019-12-18 14:26:18.136163+00    761.0  2019-12-18 14:26:18.128407+00   
4   2020-05-07 09:21:55.34008+00   7686.0  2020-05-07 09:21:55.320193+00   

   deleted_account_id             reimbursement_date  \
0                 NaN  2020-01-09 19:05:21.596363+00   
1                 NaN   2020-01-09 19:50:12.34778+00   
2                 NaN  2020-01-09 19:13:35.825041+00   
3                 NaN  2020-01-09 19:16:10.879606+00   
4                 NaN         2020-06-05 22:00:00+00   

  cash_request_received_date money_back_date transfer_type send_at  \
0                        NaN             NaN       regular     NaN   
1                        NaN             NaN       regular     NaN   
2                        NaN             NaN       regular     NaN   
3                        NaN             NaN       regular     NaN   
4                        NaN             NaN       regular     NaN   

  recovery_status reco_creation reco_last_update  
0             NaN           NaN              NaN  
1             NaN           NaN              NaN  
2             NaN           NaN              NaN  
3             NaN           NaN              NaN  
4             NaN           NaN              NaN

In [97]:
#looking at the numerical values
display(fees.describe().T, cash.describe().T)

count          mean          std     min      25%      50%  \
id               21061.0  10645.355111  6099.315256     1.0   5385.0  10652.0   
cash_request_id  21057.0  16318.449162  6656.149949  1456.0  11745.0  17160.0   
total_amount     21061.0      5.000237     0.034453     5.0      5.0      5.0   

                     75%      max  
id               15925.0  21193.0  
cash_request_id  21796.0  27010.0  
total_amount         5.0     10.0

count          mean           std   min       25%  \
id                  23970.0  13910.966124   7788.117214   3.0   7427.25   
amount              23970.0     82.720818     26.528065   1.0     50.00   
user_id             21867.0  32581.250789  27618.565773  34.0  10804.00   
deleted_account_id   2104.0   9658.755228   7972.743249  91.0   3767.00   

                        50%       75%       max  
id                  14270.5  20607.75   27010.0  
amount                100.0    100.00     200.0  
user_id             23773.0  46965.00  103719.0  
deleted_account_id   6121.5  16345.00   30445.0

In [98]:
#looking at each data column and it's type
fees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21061 entries, 0 to 21060
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               21061 non-null  int64  
 1   cash_request_id  21057 non-null  float64
 2   type             21061 non-null  object 
 3   status           21061 non-null  object 
 4   category         2196 non-null   object 
 5   total_amount     21061 non-null  float64
 6   reason           21061 non-null  object 
 7   created_at       21061 non-null  object 
 8   updated_at       21061 non-null  object 
 9   paid_at          15531 non-null  object 
 10  from_date        7766 non-null   object 
 11  to_date          7766 non-null   object 
 12  charge_moment    21061 non-null  object 
dtypes: float64(2), int64(1), object(10)
memory usage: 2.1+ MB


In [99]:
cash.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23970 entries, 0 to 23969
Data columns (total 16 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          23970 non-null  int64  
 1   amount                      23970 non-null  float64
 2   status                      23970 non-null  object 
 3   created_at                  23970 non-null  object 
 4   updated_at                  23970 non-null  object 
 5   user_id                     21867 non-null  float64
 6   moderated_at                16035 non-null  object 
 7   deleted_account_id          2104 non-null   float64
 8   reimbursement_date          23970 non-null  object 
 9   cash_request_received_date  16289 non-null  object 
 10  money_back_date             16543 non-null  object 
 11  transfer_type               23970 non-null  object 
 12  send_at                     16641 non-null  object 
 13  recovery_status             333

### Data Quality Analysis

In [100]:
# checking for null in fees
fees.isnull().sum()

id                     0
cash_request_id        4
type                   0
status                 0
category           18865
total_amount           0
reason                 0
created_at             0
updated_at             0
paid_at             5530
from_date          13295
to_date            13295
charge_moment          0
dtype: int64

There are four null in cash request id but so few we think it will not disturb the analysis, as null elements cannot be used in joining

In [101]:
#checking for null cash requests
cash.isnull().sum()

id                                0
amount                            0
status                            0
created_at                        0
updated_at                        0
user_id                        2103
moderated_at                   7935
deleted_account_id            21866
reimbursement_date                0
cash_request_received_date     7681
money_back_date                7427
transfer_type                     0
send_at                        7329
recovery_status               20640
reco_creation                 20640
reco_last_update              20640
dtype: int64

There are 2103 null values for user_id. Those would need to be excluded from a frequency analysis.

In [102]:
# dropping users with NaN user_id
cash.dropna(subset=["user_id"], inplace=True)
cash["user_id"].isnull().sum()

0

### Creating the Columns for the Cohorts

In [103]:
# creating a column with YYYY-MM-DD for the activity month. 
# activity month is the month is which the cash request was made

cash["activity_month"] = cash["created_at"].str[:7] + "-01"

In [104]:
cash[["id", "created_at", "activity_month"]]

id                     created_at activity_month
0          5  2019-12-10 19:05:21.596873+00     2019-12-01
1         70   2019-12-10 19:50:12.34778+00     2019-12-01
2          7   2019-12-10 19:13:35.82546+00     2019-12-01
3         10  2019-12-10 19:16:10.880172+00     2019-12-01
4       1594  2020-05-06 09:59:38.877376+00     2020-05-01
...      ...                            ...            ...
23964  16834  2020-09-19 07:37:14.351567+00     2020-09-01
23965  20616  2020-10-12 13:54:11.686225+00     2020-10-01
23967  22357  2020-10-20 07:58:04.006937+00     2020-10-01
23968  20256  2020-10-10 05:40:55.700422+00     2020-10-01
23969  19886  2020-10-08 14:16:52.155661+00     2020-10-01

[21867 rows x 3 columns]

In [105]:
#checking the dtype of activity month
cash[["id", "created_at", "activity_month"]].dtypes

id                 int64
created_at        object
activity_month    object
dtype: object

In [106]:
# making sure that we have a datetime object for activity month
cash["activity_month"] = cash["activity_month"].astype("datetime64[s]")

In [107]:
#checking activity month has changed dtype to datetime
cash[["id", "created_at", "activity_month"]].dtypes

id                        int64
created_at               object
activity_month    datetime64[s]
dtype: object

In [108]:
# creating cohorts
# users are assigned to the cohort of their first activity.
cash["cohort_month"] = cash["user_id"].map(cash.groupby("user_id")["activity_month"].min()) 

In [109]:
# adding a new column with cohort as index
# looking to group the activites to the month after the cohort started
cash["cohort_index"] = (cash["activity_month"].dt.to_period("M").astype(int) \
                           - cash["cohort_month"].dt.to_period("M").astype(int))

In [110]:
# looking at the columns needed
cash[["id", "created_at", "user_id", "activity_month", "cohort_month", "cohort_index"]]

id                     created_at  user_id activity_month  \
0          5  2019-12-10 19:05:21.596873+00    804.0     2019-12-01   
1         70   2019-12-10 19:50:12.34778+00    231.0     2019-12-01   
2          7   2019-12-10 19:13:35.82546+00    191.0     2019-12-01   
3         10  2019-12-10 19:16:10.880172+00    761.0     2019-12-01   
4       1594  2020-05-06 09:59:38.877376+00   7686.0     2020-05-01   
...      ...                            ...      ...            ...   
23964  16834  2020-09-19 07:37:14.351567+00  20064.0     2020-09-01   
23965  20616  2020-10-12 13:54:11.686225+00  13681.0     2020-10-01   
23967  22357  2020-10-20 07:58:04.006937+00  82122.0     2020-10-01   
23968  20256  2020-10-10 05:40:55.700422+00  64517.0     2020-10-01   
23969  19886  2020-10-08 14:16:52.155661+00  44867.0     2020-10-01   

      cohort_month  cohort_index  
0       2019-12-01             0  
1       2019-12-01             0  
2       2019-12-01             0  
3       2019-12-01             0  
4       2020-05-01             0  
...            ...           ...  
23964   2020-06-01             3  
23965   2020-06-01             4  
23967   2020-10-01             0  
23968   2020-10-01             0  
23969   2020-09-01             1  

[21867 rows x 6 columns]

In [111]:
#find a user with the maximum frequency of cash requests
cash["user_id"].value_counts()

user_id
3377.0     19
13851.0    18
2142.0     18
2530.0     17
10086.0    17
           ..
64529.0     1
68890.0     1
53757.0     1
83075.0     1
89669.0     1
Name: count, Length: 10798, dtype: int64

In [112]:
# creating a dataframe for the user with the maximum cash requests
user_max = cash[cash["user_id"] == 3377][["id", "created_at", "user_id", "activity_month", "cohort_month", "cohort_index"]]

In [113]:
# looking at the data for that user sorted in descending order
user_max.sort_values(by='activity_month', ascending = False)

id                     created_at  user_id activity_month  \
23960  26969  2020-11-01 17:31:03.345533+00   3377.0     2020-11-01   
5416   14530    2020-09-04 13:31:11.0583+00   3377.0     2020-09-01   
1664   11450  2020-08-08 06:57:10.590228+00   3377.0     2020-08-01   
5206   12630  2020-08-15 13:23:34.375488+00   3377.0     2020-08-01   
4385    9231  2020-07-21 10:54:52.748331+00   3377.0     2020-07-01   
4358   10266  2020-07-31 13:12:50.911871+00   3377.0     2020-07-01   
2764    7832  2020-07-13 15:22:09.832988+00   3377.0     2020-07-01   
1920    6143   2020-07-01 14:19:01.30501+00   3377.0     2020-07-01   
1705    3642  2020-06-17 13:56:27.141261+00   3377.0     2020-06-01   
1582    2819  2020-06-09 18:36:00.336349+00   3377.0     2020-06-01   
840     2036  2020-05-19 15:52:55.868692+00   3377.0     2020-05-01   
614     1684  2020-05-11 17:08:36.422278+00   3377.0     2020-05-01   
408     2243  2020-05-28 14:39:45.591884+00   3377.0     2020-05-01   
30      1540  2020-05-04 04:23:57.148763+00   3377.0     2020-05-01   
729     1019  2020-04-01 16:26:07.710169+00   3377.0     2020-04-01   
251     1369  2020-04-24 18:51:10.588402+00   3377.0     2020-04-01   
229      936  2020-03-18 08:05:49.092391+00   3377.0     2020-03-01   
443      987  2020-03-25 09:12:17.659574+00   3377.0     2020-03-01   
230      830  2020-03-09 13:31:06.855375+00   3377.0     2020-03-01   

      cohort_month  cohort_index  
23960   2020-03-01             8  
5416    2020-03-01             6  
1664    2020-03-01             5  
5206    2020-03-01             5  
4385    2020-03-01             4  
4358    2020-03-01             4  
2764    2020-03-01             4  
1920    2020-03-01             4  
1705    2020-03-01             3  
1582    2020-03-01             3  
840     2020-03-01             2  
614     2020-03-01             2  
408     2020-03-01             2  
30      2020-03-01             2  
729     2020-03-01             1  
251     2020-03-01             1  
229     2020-03-01             0  
443     2020-03-01             0  
230     2020-03-01             0

As we can see the longer the activity period is the larger the cohort index will be.

In [114]:
# checking that index also works with more than a year long window
user_max2 = cash[cash["user_id"] == 526][["id", "created_at", "user_id", "activity_month", "cohort_month", "cohort_index"]]
user_max2.sort_values(by='activity_month', ascending = False)

id                     created_at  user_id activity_month  \
8373   20625  2020-10-12 14:23:15.276958+00    526.0     2020-10-01   
21451  20665  2020-10-12 17:22:01.397186+00    526.0     2020-10-01   
21620  26720  2020-10-31 12:46:40.831215+00    526.0     2020-10-01   
6528   15856  2020-09-12 12:32:13.702831+00    526.0     2020-09-01   
6573   16070    2020-09-14 13:13:07.2519+00    526.0     2020-09-01   
6631   15995  2020-09-13 22:54:23.288454+00    526.0     2020-09-01   
6654   16025  2020-09-14 09:09:50.959096+00    526.0     2020-09-01   
14483  12484  2020-08-14 12:28:42.222162+00    526.0     2020-08-01   
12476   7250  2020-07-08 22:46:36.854214+00    526.0     2020-07-01   
9533    3478  2020-06-16 12:38:38.140763+00    526.0     2020-06-01   
9165    1348  2020-04-23 18:54:05.810157+00    526.0     2020-04-01   
7869     824   2020-03-08 14:01:05.14633+00    526.0     2020-03-01   
9194     985  2020-03-25 02:27:41.780578+00    526.0     2020-03-01   
5822     614  2020-02-05 17:54:50.727851+00    526.0     2020-02-01   
7971     131  2019-12-12 20:59:48.763947+00    526.0     2019-12-01   

      cohort_month  cohort_index  
8373    2019-12-01            10  
21451   2019-12-01            10  
21620   2019-12-01            10  
6528    2019-12-01             9  
6573    2019-12-01             9  
6631    2019-12-01             9  
6654    2019-12-01             9  
14483   2019-12-01             8  
12476   2019-12-01             7  
9533    2019-12-01             6  
9165    2019-12-01             4  
7869    2019-12-01             3  
9194    2019-12-01             3  
5822    2019-12-01             2  
7971    2019-12-01             0

### Creating the Frequency Counts
Frequency of Service Usage: Understand how often users from each cohort utilize IronHack Payments' cash advance services over time. For this analysis we don't need the fees dataset

### Merging Fees and Cash Requests
The id from the cash request file matches with cash_request_id from the fees file.

In [115]:
#checking how many cash_request_ids from fees match with id from cash
matching_ids = cash[cash['id'].isin(fees['cash_request_id'])].shape[0]
#checking how many cash_request_ids do not match with id from cash
missing_ids_in_fees = cash[~cash['id'].isin(fees['cash_request_id'])].shape[0]
print(f"There are {matching_ids} matching cash request ids in fees. In comparison, there is a total \
number of {cash["id"].shape[0]} cash request ids. This makes a difference of {missing_ids_in_fees}")

There are 12496 matching cash request ids in fees. In comparison, there is a total number of 21867 cash request ids. This makes a difference of 9371


The reason for more ids in cash compared to cash_request_id in fees might lie in the fact that a lot of cash requests got rejected or have not been paid back yet (no fees yet)

In [116]:
cash["status"].value_counts()

status
money_back               15683
rejected                  5221
direct_debit_rejected      815
active                      59
transaction_declined        48
direct_debit_sent           34
canceled                     7
Name: count, dtype: int64

### Revenue Generated by the Cohort: 
Calculate the total revenue generated by each cohort over months to assess the financial impact of user behavior.

In [117]:
#total fees by cash_request_id for all accepted payments
fees_agg_revenue = fees[fees["status"] == 'accepted'].groupby("cash_request_id")[["total_amount"]].sum()
fees_agg_revenue.reset_index()

cash_request_id  total_amount
0               1518.0           5.0
1               1554.0          10.0
2               1556.0           5.0
3               1566.0           5.0
4               1625.0           5.0
...                ...           ...
11841          27004.0           5.0
11842          27006.0           5.0
11843          27007.0           5.0
11844          27009.0           5.0
11845          27010.0           5.0

[11846 rows x 2 columns]

In [119]:
merged = cash.merge(fees_agg_revenue, how ='left', left_on='id', right_on='cash_request_id')
merged

id  amount             status                     created_at  \
0          5   100.0           rejected  2019-12-10 19:05:21.596873+00   
1         70   100.0           rejected   2019-12-10 19:50:12.34778+00   
2          7   100.0           rejected   2019-12-10 19:13:35.82546+00   
3         10    99.0           rejected  2019-12-10 19:16:10.880172+00   
4       1594   100.0           rejected  2020-05-06 09:59:38.877376+00   
...      ...     ...                ...                            ...   
21862  16834   100.0         money_back  2020-09-19 07:37:14.351567+00   
21863  20616   100.0         money_back  2020-10-12 13:54:11.686225+00   
21864  22357   100.0         money_back  2020-10-20 07:58:04.006937+00   
21865  20256   100.0         money_back  2020-10-10 05:40:55.700422+00   
21866  19886   100.0  direct_debit_sent  2020-10-08 14:16:52.155661+00   

                          updated_at  user_id                   moderated_at  \
0       2019-12-11 16:47:42.40783+00    804.0  2019-12-11 16:47:42.405646+00   
1      2019-12-11 14:24:22.900054+00    231.0  2019-12-11 14:24:22.897988+00   
2      2019-12-11 09:46:59.779773+00    191.0  2019-12-11 09:46:59.777728+00   
3      2019-12-18 14:26:18.136163+00    761.0  2019-12-18 14:26:18.128407+00   
4       2020-05-07 09:21:55.34008+00   7686.0  2020-05-07 09:21:55.320193+00   
...                              ...      ...                            ...   
21862  2021-02-05 12:11:58.186458+00  20064.0  2020-09-19 15:37:54.670099+00   
21863  2021-02-06 20:17:49.292493+00  13681.0                            NaN   
21864  2021-02-05 12:19:30.656816+00  82122.0                            NaN   
21865  2021-02-05 13:14:19.707627+00  64517.0                            NaN   
21866  2021-01-05 15:45:52.645536+00  44867.0                            NaN   

       deleted_account_id             reimbursement_date  \
0                     NaN  2020-01-09 19:05:21.596363+00   
1                     NaN   2020-01-09 19:50:12.34778+00   
2                     NaN  2020-01-09 19:13:35.825041+00   
3                     NaN  2020-01-09 19:16:10.879606+00   
4                     NaN         2020-06-05 22:00:00+00   
...                   ...                            ...   
21862                 NaN         2020-10-05 22:00:00+00   
21863                 NaN         2021-02-06 11:00:00+00   
21864                 NaN         2021-02-05 11:00:00+00   
21865                 NaN         2021-02-05 11:00:00+00   
21866                 NaN         2021-02-05 11:00:00+00   

      cash_request_received_date                money_back_date transfer_type  \
0                            NaN                            NaN       regular   
1                            NaN                            NaN       regular   
2                            NaN                            NaN       regular   
3                            NaN                            NaN       regular   
4                            NaN                            NaN       regular   
...                          ...                            ...           ...   
21862                 2020-09-22  2021-02-05 12:11:58.131036+00       instant   
21863                 2020-10-13  2021-02-06 20:17:49.257521+00       instant   
21864                 2020-10-21  2021-02-05 12:19:30.626289+00       instant   
21865                 2020-10-12  2021-02-05 13:14:19.689906+00       instant   
21866                 2020-10-10                            NaN       instant   

                             send_at recovery_status  \
0                                NaN             NaN   
1                                NaN             NaN   
2                                NaN             NaN   
3                                NaN             NaN   
4                                NaN             NaN   
...                              ...             ...   
21862  2020-09-26 07:37:14.299162+00       completed   
21863  2020-10-12 

In [ ]:
fees_agg_revenue.columns

AttributeError: 'Series' object has no attribute 'columns'